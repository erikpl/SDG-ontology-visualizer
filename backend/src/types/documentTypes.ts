
export type Document = {
    celexID: string;
    id: string;
    title: string;
    language: LanguageReference;
    format: string;
    url: string;
}

export type Language = {
    isoCode: string;
    // What goes here?
}

export type DocumentSDGRelation = {
    celexID: string;
    goals: string[];
    targets: Map<string, string[]>;
}

export type LanguageReference = string;