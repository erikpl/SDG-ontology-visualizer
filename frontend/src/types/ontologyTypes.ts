import { SimulationNodeDatum } from 'd3';

export type Annotation = {
  label: string;
  description: string;
  moreInformation: string | null;
};

export type SustainabilityGoal = {
  instancesOf: string;
  label: string;
  icon: string;
};

export interface SubGoal extends Node {
  SubjectLabel: string;
  description: string;
}

export interface UniqueObject {
  id: string;
}

export interface Node extends UniqueObject {
  name: string;
  type: string;
  prefix: Prefix;
  correlation: number;
}

export interface GraphNode extends Node, SimulationNodeDatum {
  isLocked?: boolean;
  lockedX?: number;
  lockedY?: number;
  color?: string;
  type: string;
}

export interface Edge extends UniqueObject {
  name: string;
  prefix: Prefix;
  correlation: number;
}

export interface GraphEdge extends Edge {
  source: string;
  target: string;
  sourceToTarget: Edge[];
  targetToSource: Edge[];
}

export type Ontology = {
  Subject: Node;
  Object: Node;
  Predicate: Edge;
};

export type Prefix = {
  prefix: string;
  iri: string;
};

export type CorrelationFilter = {
  pLow: boolean;
  pMedium: boolean;
  pHigh: boolean;
  nLow: boolean;
  nMedium: boolean;
  nHigh: boolean;
};

export type Document = {
  celexID: string;
  title: string;
  language: string;
  format: string;
  url: string;
};

export type LanguageItem = {
  id: number;
  ISO_639_1: ISO6391Code;
  ISO_639_2T: ISO6392TCode;
};

export enum ISO6391Code {
  bg, cs, da, de, el, en, es, et, fi, fr, ga, 
  hr, hu, it, lt, lv, mt, nl, no, pl, pt, ro, sk, sl, sv, null,
}

export enum ISO6392TCode {
  HRV,
  FIN,
  CES,
  SPA,
  DAN,
  EST,
  POL,
  DEU,
  ELL,
  POR,
  SLV,
  LAV,
  LIT,
  BUL,
  ITA,
  GLE,
  RON,
  HUN,
  SWE,
  MLT,
  ENG,
  SLK,
  FRA,
  NOR,
  NLD,
  null,
}