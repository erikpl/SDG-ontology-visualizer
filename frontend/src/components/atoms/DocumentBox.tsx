/* eslint-disable react/jsx-props-no-spreading */
import { CheckCircleIcon } from '@chakra-ui/icons';
import {
  Accordion,
  AccordionButton,
  AccordionIcon,
  AccordionItem,
  AccordionPanel,
  Badge,
  Button,
  Flex,
  Heading,
  HStack,
  Stack,
  Text,
  VStack,
} from '@chakra-ui/react';
import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';
import { getRelatedSubgoalsForDocument, getSustainabilityGoal } from '../../api/ontologies';
import useTranslation from '../../hooks/translations';
import { selectSDG, selectSubgoal } from '../../state/reducers/ontologyReducer';
import { Document, SubGoal } from '../../types/ontologyTypes';
import flagComponentsAlt from '../../utils/flagUtilsAlt';

type DocumentBoxProps = {
  commonCelexDocuments: Document[][];
};

const DocumentBox: React.FC<DocumentBoxProps> = ({ 
  commonCelexDocuments, 
}: DocumentBoxProps) => {
  const [relatedGoals, setRelatedGoals] = useState<Array<Array<SubGoal>>>([]);
  const history = useHistory();
  const dispatch = useDispatch();
  const translations = useTranslation(); 

  const onClickSeeRelatedGoals = async (document: Document) => {
    if (relatedGoals.length === 0){
      const goals = await getRelatedSubgoalsForDocument(document.url);
      setRelatedGoals(goals);
    }
  };

  const getGoalLabelFromSubgoal = (subgoal: SubGoal) => subgoal.SubjectLabel.slice(0, subgoal.SubjectLabel.indexOf('.'));
  
  const formatTitle = (title: string) => {
    const tokens = title.split(' ');
    console.log(title);
    if (tokens.length > 50) {
      return tokens.slice(0, 50).join(' ').concat('...');
    }
    return title;
  };

  const onClickSubGoal = async (subgoal: SubGoal) => {
    const goal = await getSustainabilityGoal(getGoalLabelFromSubgoal(subgoal));
    dispatch(selectSDG(goal[0]));
    dispatch(selectSubgoal(subgoal));
  };

  const Flag = (document: Document) => {
    const { language } = document;
    const CountryFlag = flagComponentsAlt[language];
    return <CountryFlag />;
  };

  return (
    <Accordion allowMultiple allowToggle width='80%' onChange={() => onClickSeeRelatedGoals(commonCelexDocuments[0][0])} backgroundColor='white'>
      <AccordionItem boxShadow='lg' borderWidth='2px' borderRadius='md' borderColor='cyan.700'>
        <AccordionButton borderRadius='md' _hover={{ opacity: '75%' }}>
          <HStack> 
            <Badge width='40px' padding='5px' backgroundColor='cyan.100' marginRight='30px'>
              <Flag {...commonCelexDocuments[0][0]} />
            </Badge>  
            <Heading as='h3' size='sm' width='90%' textAlign='justify' fontWeight='medium' marginRight='30px'>
              {formatTitle(commonCelexDocuments[0][0].title)}
            </Heading>
          </HStack>
        
          <AccordionIcon />
        </AccordionButton>
        <AccordionPanel>
          <Stack spacing='5'>
            <Text size='md'>
              <CheckCircleIcon color='green.400' />
              &nbsp;
              {translations.getString('Active')}
            </Text>
            <Flex justify='space-evenly'>
              {commonCelexDocuments[0].map(document => (
                <Button bg='cyan.700' color='white' _hover={{ opacity: '75%' }} key={document.url}>
                  <a href={document.url} target='_blank' rel='noreferrer'>
                    {translations.getString('OpenAs')}
                    &nbsp;
                    {document.format}
                  </a>
                </Button>
              ))}
            </Flex>
            <AccordionItem>
              <AccordionButton>
                {translations.getString('SeeRelatedGoals')}
                <AccordionIcon /> 
              </AccordionButton>
              <AccordionPanel>
                <VStack align='start' justify='center'> 
                  {relatedGoals.map(subgoalList => (
                    <HStack align='start'>
                      <Text width='20px' textAlign='end' fontWeight='bold'>{getGoalLabelFromSubgoal(subgoalList[0])}</Text>
                      {subgoalList.map(subgoal => (
                        <Button 
                          bgColor='cyan.100' 
                          width='40px' 
                          height='20px'
                          onClick={() => {
                          window.scrollTo(0, 0);
                          onClickSubGoal(subgoal);
                          history.push('/documents');
                        }}
                        >
                          {subgoal.SubjectLabel}
                        </Button>
                        ))}
                    </HStack>
                  ))}
                </VStack>
              </AccordionPanel>
            </AccordionItem>
          
          </Stack>
        </AccordionPanel>
      </AccordionItem>
    </Accordion>
  );
};


export default DocumentBox;
