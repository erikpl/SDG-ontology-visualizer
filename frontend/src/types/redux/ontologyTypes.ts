import { CorrelationFilter, Node, SubGoal, SustainabilityGoal } from '../ontologyTypes';

export type SetCorrelationFilterPayload = {
  isPositive: boolean;
  index: number;
};

export type OntologyState = {
  selectedNode?: Node;
  selectedSDG?: SustainabilityGoal;
  selectedSubGoal?: SubGoal;
  correlationFilter: CorrelationFilter;
};

export type SelectNodeAction = {
  type: typeof SELECT_NODE;
  payload: Node;
};

export type ClearSelectedNodeAction = {
  type: typeof CLEAR_SELECTED_NODE;
};

export type SetCorrelationFilterAction = {
  type: typeof SET_CORRELATION_FILTER;
  payload: SetCorrelationFilterPayload;
};

export type SelectSDGAction = {
  type: typeof SELECT_SDG;
  payload: SustainabilityGoal;
};

export type SelectSubgoalAction = {
  type: typeof SELECT_SUBGOAL;
  payload: SubGoal;
};

export type OntologyStateAction =
  | SelectNodeAction
  | ClearSelectedNodeAction
  | SetCorrelationFilterAction
  | SelectSDGAction
  | SelectSubgoalAction;

export const SELECT_NODE = 'SELECT_NODE';
export const SELECT_SDG = 'SELECT_SDG';
export const SELECT_SUBGOAL = 'SELECT_SUBGOAL';
export const CLEAR_SELECTED_NODE = 'CLEAR_SELECTED_NODE';
export const SET_CORRELATION_FILTER = 'SET_CORRELATION_FILTER';
