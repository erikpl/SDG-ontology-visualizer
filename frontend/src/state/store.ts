import { combineReducers, createStore } from 'redux';
import apiErrorReducer from './reducers/apiErrorReducer';
import ontologyReducer from './reducers/ontologyReducer';
import fullscreenReducer from './reducers/fullscreenReducer';
import loginReducer from './reducers/loginReducer';
import languageReducer from './reducers/languageReducer';

const rootReducer = combineReducers({
  apiError: apiErrorReducer,
  ontology: ontologyReducer,
  fullscreenStatus: fullscreenReducer,
  login: loginReducer,
  languages: languageReducer,
});

const store = createStore(rootReducer);

export type RootState = ReturnType<typeof rootReducer>;

export default store;
