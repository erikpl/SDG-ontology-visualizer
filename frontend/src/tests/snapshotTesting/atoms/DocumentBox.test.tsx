import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import store from '../../../state/store';

import DocumentBox from '../../../components/atoms/DocumentBox';
import { Document } from '../../../types/ontologyTypes';
import LanguagesList from '../../../localization/languages';

const testDocs: Document[][] = [];
const testFormats = ['pdfa2a', 'fmx4', 'xhtml'];
const testCelex = 'testCelex';
const testTitle = 'testTitle';
const testUrl = 'testUrl';

LanguagesList.map((language) => {
    let commonLangDocs: Document[] = [];
    testFormats.map((format) => {
        let doc: Document = {
            celexID: testCelex,
            title: testTitle,
            language: language.ISO_639_2T.toString(),
            format: format,
            url: testUrl
        }
        commonLangDocs.push(doc);
    });
    testDocs.push(commonLangDocs);
});

it('renders with a list of dummy docs', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <DocumentBox commonCelexDocuments={testDocs} />
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});