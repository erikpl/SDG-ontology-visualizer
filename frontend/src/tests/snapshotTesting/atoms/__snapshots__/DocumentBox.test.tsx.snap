// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with a list of dummy docs 1`] = `
<div
  className="chakra-accordion css-1x97c6v"
>
  <div
    className="chakra-accordion__item css-16blszo"
  >
    <button
      aria-controls="accordion-panel-2"
      aria-expanded={false}
      className="chakra-accordion__button css-1ne4ibq"
      disabled={false}
      id="accordion-button-2"
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      type="button"
    >
      <h3
        className="chakra-heading css-0"
      >
        testTitle
      </h3>
      <span
        className="chakra-badge css-ifsvay"
      >
        13
      </span>
      <svg
        aria-hidden={true}
        className="chakra-icon chakra-accordion__icon css-8o5ky2"
        focusable={false}
        viewBox="0 0 24 24"
      >
        <path
          d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          fill="currentColor"
        />
      </svg>
    </button>
    <div
      className="chakra-collapse"
      style={
        Object {
          "display": "none",
          "height": "0px",
          "opacity": 0,
          "overflow": "hidden",
        }
      }
    >
      <div
        aria-labelledby="accordion-button-2"
        className="chakra-accordion__panel css-0"
        id="accordion-panel-2"
        role="region"
      >
        <div
          className="chakra-stack css-1eitvxc"
        >
          <p
            className="chakra-text css-0"
          >
            <svg
              className="chakra-icon css-m17ft2"
              focusable={false}
              viewBox="0 0 24 24"
            >
              <path
                d="M12,0A12,12,0,1,0,24,12,12.014,12.014,0,0,0,12,0Zm6.927,8.2-6.845,9.289a1.011,1.011,0,0,1-1.43.188L5.764,13.769a1,1,0,1,1,1.25-1.562l4.076,3.261,6.227-8.451A1,1,0,1,1,18.927,8.2Z"
                fill="currentColor"
              />
            </svg>
             Aktiv
          </p>
          <div
            className="css-102lqhp"
          >
            <button
              className="chakra-button css-gwt898"
              type="button"
            >
              <a
                href="testUrl"
                rel="noreferrer"
                target="_blank"
              >
                Åpne som 
                pdfa2a
              </a>
            </button>
            <button
              className="chakra-button css-gwt898"
              type="button"
            >
              <a
                href="testUrl"
                rel="noreferrer"
                target="_blank"
              >
                Åpne som 
                fmx4
              </a>
            </button>
            <button
              className="chakra-button css-gwt898"
              type="button"
            >
              <a
                href="testUrl"
                rel="noreferrer"
                target="_blank"
              >
                Åpne som 
                xhtml
              </a>
            </button>
          </div>
          <div
            className="chakra-accordion__item css-1fsnuue"
          >
            <button
              aria-controls="accordion-panel-1"
              aria-expanded={false}
              className="chakra-accordion__button css-9179w8"
              disabled={false}
              id="accordion-button-1"
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              type="button"
            >
              Se relaterte mål
              <svg
                aria-hidden={true}
                className="chakra-icon chakra-accordion__icon css-8o5ky2"
                focusable={false}
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  fill="currentColor"
                />
              </svg>
            </button>
            <div
              className="chakra-collapse"
              style={
                Object {
                  "display": "none",
                  "height": "0px",
                  "opacity": 0,
                  "overflow": "hidden",
                }
              }
            >
              <div
                aria-labelledby="accordion-button-1"
                className="chakra-accordion__panel css-0"
                id="accordion-panel-1"
                role="region"
              >
                <div
                  className="chakra-stack css-m203cn"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
